swarm/storage/ldbstore.go

```diff
@@ -74,8 +74,33 @@ type gcItem struct {
 	idxKey []byte
 }
 
+type LDBStoreParams struct {
+	*StoreParams
+	Path string
+	Po   func(Key) uint8
+}
+
+// create params with specified values.
+// If 0 for capacity or nil for hash, baskey is specified, default values for these params will be used
+// path has no default value
+func NewLDBStoreParams(path string, capacity uint64, hash SwarmHasher, basekey []byte) *LDBStoreParams {
+	if hash == nil {
```

Comment:
I think we should avoid doing this. Modifying the behaviour of constructors, by passing `nil` values or booleans is not very intuitive, and I have already been beaten by this in the `simulation` framework code, which I refactored.

We should be explicit with the parameters we use for constructors, or just have different constructors.

If we have a number of parameters that take different values, we could start of with a default set, and then introduce chaining of parameters for example.

In this particular example, it is impossible to deduce what the constructor does, without actually reading it, so I'd make the following two changes:
1. Introduce a default constructor which doesn't take a `hash` and always initialises the `LDBStoreParams` with `SHA3Hash`, and also always initialises `capacity` to `singletonSwarmDbCapacity`.
2. Then add 2 chained methods, `WithCapacity`, `WithSwarmHasher`, that actually modify these default values.

Let me know this makes sense of if you disagree :)

Reply:
I am for `With*` functions for optional parameters for constructors and functions in general. They have a disadvantage to have a longer names to avoid collisions in bigger packages, but there is a method to avoid that and preserve the type safety, but require more glue code. I believe that this should be addressed in during the rewrite of storage package.